Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaFillAlgorithm.kt

```diff
@@ -0,0 +1,184 @@
+        )
+        ranges.add(range)
+        ranges.add(Range(range.line, range.start, range.end, DOWN))
+        var row: Int
+        while (!ranges.isEmpty()) {
+            range = ranges.poll() ?: break
+            val direction = range.direction
+            val diff = if (direction == UP) -1 else 1
+            row = range.line + diff
+            if (row in 0 until height) {
+                checkRangeAndGenerateNewRanges(range, row, direction)
+            }
+        }
+    }
+
+    private fun square(x: Int) = x * x
+
+    private fun shouldCellBeFilled(row: Int, col: Int): Boolean =
+        !filledPixels[row][col] && (
+            pixels[row][col] == replacementColor || considerTolerance && isPixelWithinColorTolerance(
+                pixels[row][col],
+                replacementColor
+            )
+            )
+
+    private fun validateAndAssign(row: Int, col: Int): Boolean = if (shouldCellBeFilled(row, col)) {
+        pixels[row][col] = targetColor
+        filledPixels[row][col] = true
+        true
+    } else false
+
+    private fun getStartIndex(row: Int, col: Int): Int {
+        val start = (col downTo 0).find { !validateAndAssign(row, it) } ?: -1
+        return start + 1
+    }
+
+    private fun getEndIndex(row: Int, col: Int): Int {
+        val end = (col until width).find { !validateAndAssign(row, it) } ?: width
+        return end - 1
+    }
+
+    private fun generateRangeAndReplaceColor(row: Int, col: Int, direction: Boolean): Range {
+        val range = Range()
+        pixels[row][col] = targetColor
+        filledPixels[row][col] = true
+
+        val start = getStartIndex(row, col - 1)
+        val end = getEndIndex(row, col + 1)
+        range.apply {
+            line = row
+            this.end = end
+            this.start = start
+            this.direction = direction
+        }
+        bitmap.setPixels(pixels[row], start, width, start, row, end - start + 1, 1)
+        return range
+    }
+
+    private fun checkRangeAndGenerateNewRanges(range: Range, row: Int, directionUp: Boolean) {
+        var newRange: Range
+        var col = range.start
+        while (col <= range.end) {
+            if (shouldCellBeFilled(row, col)) {
+                newRange = generateRangeAndReplaceColor(row, col, directionUp)
+                ranges.add(newRange)
+                if (newRange.start <= range.start - 2) {
+                    ranges.add(Range(row, newRange.start, range.start - 2, !directionUp))
+                }
+                if (newRange.end >= range.end + 2) {
+                    ranges.add(Range(row, range.end + 2, newRange.end, !directionUp))
+                }
+                col = if (newRange.end >= range.end - 1) {
+                    break
+                } else {
+                    newRange.end + 1
+                }
+            }
+            col++
+        }
+    }
+
+    private fun isPixelWithinColorTolerance(pixel: Int, referenceColor: Int): Boolean {
```

Comment:
Hmm, I understand this method, but I know there is even a better way. We can get rid of the constants :tada: 
https://developer.android.com/reference/android/graphics/Color?hl=es#public-methods
There are static getters or use the Color class directly. One could simply override the Operators (private extension function or create a helper file)
Possible extension methods/operators for the Color class: addition, subtraction, the squared distance [with(out) alpha]
https://kotlinlang.org/spec/operator-overloading.html